### PR TITLE
netd changes to support dual stack (ipv4_ipv6)

### PR DESCRIPTION
**Pull request by basantsa1989**
When IPV6 is enabled, we'll populate the ipv6Subnet from the value set in spec.podCIDRs[1]

### CODE REVIEW

File: scripts/install-cni.sh
Comment (MrHohn) [inline scripts/install-cni.sh:131]: Which component is responsible to fill in `.spec.podCIDRs[1]`? Should the script exit and retry if this is not there yet?

File: scripts/install-cni.sh
Comment (MrHohn) [inline scripts/install-cni.sh:127]: Just double checking - does the same `node_ipv6_addr` checks make sense for dualstack? (Asking as I don't see `node_ipv6_addr` being used anymore.)

File: scripts/install-cni.sh
Comment (basantsa1989) [reply scripts/install-cni.sh:131]: I am working on the k8s changes required to populate .spec.podCIDRs[1].

The changes will be in:
https://github.com/kubernetes/legacy-cloud-providers/blob/master/gce/gce_instances.go#L495
and 
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go#L284

With those changes, we should always have a value in .spec.podCIDRs[1] when IPV6/DualStack is enabled.

File: scripts/install-cni.sh
Comment (basantsa1989) [reply scripts/install-cni.sh:127]: When enable_ipv6 is true, we need to have a valid node_ipv6_addr. So the node_ipv6_addr checks are still valid.
Just the way we get ipv6Subnet will be different.

File: scripts/install-cni.sh
Comment (sdmodi) [reply scripts/install-cni.sh:131]: Basant I was testing this and it seems like there is a race in the GCE metadata server and populating the interface IP. Basically what we end up writing in the podCIDRs is sometimes wrong and stale. For now we can continue using the interface's IP address to write the CNI config just like you had suggested. This would look like this:

s#@ipv6SubnetOptional#, [{\"subnet\": \"${node_ipv6_addr%:*:*}:ffff:0/112\"}]#g;

If it is safe we should also do the same thing for directpath. We cannot use the first /112 because then we will conflict with the node IP. When dual stack clusters are enabled, we want to use the last /112 for pods.

Comment (basantsa1989) [reply scripts/install-cni.sh:131]: @sdmodi I'll make this change and test. Thanks

Comment (basantsa1989) [reply scripts/install-cni.sh]: Done. Made the change and tested only for IPV6 Dual Stack for now. 
@varunmar @MrHohn do you see any issues with changing directpath to use last /112 block instead of first /112 block?

Comment (MrHohn) [reply scripts/install-cni.sh]: Not an expert on the directpath usage with pod - will defer to @varunmar to respond :)

File: scripts/install-cni.sh
Comment (sdmodi) [inline scripts/install-cni.sh:132]: This looks like a debug log. We should remove it.

**Review by sdmodi — APPROVED** [review]

File: scripts/install-cni.sh
Comment (basantsa1989) [reply scripts/install-cni.sh]: Done.

File: scripts/install-cni.sh
Comment (varunmar) [reply scripts/install-cni.sh]: I'm not sure  I understand the question - the first/last 112 block of the node allocation? How big is the range handed out to the node - it used to only be /112, is it bigger now?

Comment (basantsa1989) [reply scripts/install-cni.sh]: Node is assigned a /96 block. Node uses the first address in this block for its interface. So we need to pick a /112 block from it. I am picking the last block to avoid any ip conflicts. Want to check if this applies for directpath too.

Comment (basantsa1989) [reply scripts/install-cni.sh]: Made it consistent with directpath. Since the node address is :0 and the first address allocated to pods starts from :1, we should not see any conflicts.

**Comment by varunmar** [comment]
/approve cancel
/lgtm

Please squash the commits before merging.

**Pull request merged**
